### PR TITLE
DuckDb implementation

### PR DESCRIPTION
**Pull request by alexisszmundy**
WARNING : Merge KraftwerkTests branch "devDuckDb" into Master before deployment

### CODE REVIEW

File: kraftwerk-core/src/main/java/fr/insee/kraftwerk/core/utils/SqlUtils.java [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+
+
+    /**
+     * send CREATE TABLE queries into DB
+     *
+     * @param statement   DuckDB connection
+     * @param vtlBindings datasets to convert
+     * @throws SQLException if sql error
+     */
+    private static void createOutputSQLTables(Statement statement, VtlBindings vtlBindings) throws SQLException {
```
Comment (alicela) [inline kraftwerk-core/src/main/java/fr/insee/kraftwerk/core/utils/SqlUtils.java:54]: change method name => output is used for final output (csv, parquet)... createSchemaForSqlTables ?

File: kraftwerk-core/src/main/java/fr/insee/kraftwerk/core/utils/SqlUtils.java [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+
+                    //Remove last delimiter and replace by ")"
+                    createTableQuery.delete(createTableQuery.length() - 2, createTableQuery.length());
+                    createTableQuery.append(")");
+
+                    //Execute query
+                    log.debug("SQL Query : {}", createTableQuery);
+                    statement.execute(createTableQuery.toString());
+                }
+                insertDataIntoTable(statement.getConnection(), datasetName, vtlBindings.getDataset(datasetName), schemaOrder);
```
Comment (alicela) [inline kraftwerk-core/src/main/java/fr/insee/kraftwerk/core/utils/SqlUtils.java:84]: I prefer to call this method in convertVtlBindingsIntoSqlDatabase. 
First create tables
Next insert data

File: kraftwerk-core/src/main/java/fr/insee/kraftwerk/core/utils/SqlUtils.java [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+        for(Structured.Component component : structure.values()) {
+            deduplicatedColumns.add(component.getName().toLowerCase());
+        }
+
+        Map<String, VariableType> schema = new HashMap<>();
+        for (Structured.Component component : structure.values()) {
+            VariableType type = VariableType.getTypeFromJavaClass(component.getType());
+            if (type != null){
+                //If column not added yet
+                if(deduplicatedColumns.contains(component.getName().toLowerCase())){
```
Comment (alicela) [inline kraftwerk-core/src/main/java/fr/insee/kraftwerk/core/utils/SqlUtils.java:106]: structure.values is looped twice. You can use  schema.containsKey() instead

File: kraftwerk-core/src/main/java/fr/insee/kraftwerk/core/utils/SqlUtils.java [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+            }
+        }
+        return schema;
+    }
+
+    /**
+     * Extract columns order from a VTL dataset
+     * @return a list of columns names in the right order
+     */
+    private static List<String> extractColumnsOrder(Dataset dataset){
```
Comment (alicela) [inline kraftwerk-core/src/main/java/fr/insee/kraftwerk/core/utils/SqlUtils.java:121]: use a LinkedHashMap  for SqlSchema instead ?

File: kraftwerk-core/src/main/java/fr/insee/kraftwerk/core/outputs/csv/CsvOutputFiles.java [hunk 5]
```diff
@@ -40,17 +53,110 @@ public CsvOutputFiles(Path outDirectory, VtlBindings vtlBindings, List<String> m
...
+				//Get boolean columns names
+				List<String> boolColumnNames = SqlUtils.getColumnNames(getDatabase(), datasetName, VariableType.BOOLEAN);
+				//Get indexes of boolean columns
+				List<Integer> boolColumnIndexes = new ArrayList<>();
+
+				//Create file with double quotes header
+				Files.write(outputFile.toPath(), buildHeader(columnNames, boolColumnNames, boolColumnIndexes).getBytes());
+
+				//Data export into temp file
+				StringBuilder exportCsvQuery = new StringBuilder(String.format("COPY %s TO '%s' (FORMAT CSV, HEADER false, DELIMITER '%s', OVERWRITE_OR_IGNORE true", datasetName, outputFile.getAbsolutePath() +"data", Constants.CSV_OUTPUTS_SEPARATOR));
```
Comment (alicela) [inline kraftwerk-core/src/main/java/fr/insee/kraftwerk/core/outputs/csv/CsvOutputFiles.java:77]: extract getExportCsvQuery in a method

File: kraftwerk-core/src/main/java/fr/insee/kraftwerk/core/outputs/csv/CsvOutputFiles.java [hunk 6]
```diff
@@ -40,17 +53,110 @@ public CsvOutputFiles(Path outDirectory, VtlBindings vtlBindings, List<String> m
...
+						Files.write(outputFile.toPath(),(line + "\n").getBytes(),StandardOpenOption.APPEND);
+						line = bufferedReader.readLine();
+					}
+				}
+				Files.deleteIfExists(Path.of(outputFile.toPath().toAbsolutePath() + "data"));
+
+				//Count rows for functional log
+				if (kraftwerkExecutionLog != null) {
+					try(ResultSet countResult = this.getDatabase().executeQuery("SELECT COUNT(*) FROM " + datasetName)){
+                        assert kraftwerkExecutionLog != null; // Assert because IDE warning
```
Comment (alicela) [inline kraftwerk-core/src/main/java/fr/insee/kraftwerk/core/outputs/csv/CsvOutputFiles.java:108]: I prefer having an idiot warning than having code like that :-D

File: kraftwerk-core/src/main/java/fr/insee/kraftwerk/core/sequence/WriterSequence.java [hunk 7]
```diff
@@ -8,11 +8,13 @@
 import fr.insee.kraftwerk.core.outputs.csv.CsvOutputFiles;
```
Comment (alicela) [inline kraftwerk-core/src/main/java/fr/insee/kraftwerk/core/sequence/WriterSequence.java:8]: https://stackoverflow.com/questions/3837959/how-to-have-the-formatter-wrap-code-with-intellij

File: kraftwerk-core/src/main/java/fr/insee/kraftwerk/core/outputs/csv/CsvOutputFiles.java [hunk 8]
```diff
@@ -40,17 +53,110 @@ public CsvOutputFiles(Path outDirectory, VtlBindings vtlBindings, List<String> m
...
+		//Rebuild csv line
+		return String.join(String.valueOf(Constants.CSV_OUTPUTS_SEPARATOR),lineElements);
+	}
+
+	/**
+	 * Changes null values to "" in a line
+	 * @param csvLine line to transform
+	 * @return the transformed line
+	 */
+	private String applyNullTransformation(String csvLine) {
```
[the remark this answers is not in the document]
Comment (alexisszmundy) [reply kraftwerk-core/src/main/java/fr/insee/kraftwerk/core/outputs/csv/CsvOutputFiles.java:152]: If I try that with NULLSTR '\"\"' I have this : Binder Error: QUOTE must not appear in the NULL specification and vice versa

**Pull request merged**